src/rate_limiter.rs

```diff
@@ -813,4 +812,32 @@ mod tests {
             _ => panic!("unexpected result"),
         }
     }
+
+    /// Checks that the rate-limiter's `overcharged` container can't be over-filled.
+    ///
+    /// Keeps trying to add GET requests for `ImmutableData` with a short delay between each
+    /// attempt. Most will fail, but this should ensure the `RateLimiter` always has an entry for
+    /// this client in its `used` container (in case we go back to using the absence of a client as
+    /// a trigger to purge their overcharged entries).
+    ///
+    /// After `OVERCHARGED_TIMEOUT_SECS` plus one minute has elapsed, there should not be an
+    /// excessive number of entries in the `overcharged` container.
+    #[test]
+    fn overcharged_limit() {
+        let mut rate_limiter = RateLimiter::new(false);
+        let client = IpAddr::from([0, 0, 0, 0]);
+        let wait_millis = MAX_IMMUTABLE_DATA_SIZE_IN_BYTES * 100 / RATE as u64;
+        let max_overcharged_entries = OVERCHARGED_TIMEOUT_SECS * RATE as u64 /
+            MAX_IMMUTABLE_DATA_SIZE_IN_BYTES;
+        let finish_time = FakeClock::now() + Duration::from_secs(OVERCHARGED_TIMEOUT_SECS + 60);
+        while FakeClock::now() < finish_time {
+            let name = XorName([0; XOR_NAME_LEN]);
+            let msg_id = MessageId::new();
+            let request = UserMessage::Request(Request::GetIData { name, msg_id });
+            let request_parts = unwrap!(request.to_parts(0));
+            let _ = add_user_msg_part(&mut rate_limiter, &client, &request_parts[0]);
+            FakeClock::advance_time(wait_millis);
+        }
+        assert!((rate_limiter.overcharged.len() as u64) <= max_overcharged_entries);
```

Comment:
not sure about the config of executing this test with the previous code.
however, the current code shows that the number of entries in overcharged is maxed at the number of add_message_part calls, which is actually :
num_of_entries = (OVERCHARGED_TIMEOUT_SECS + 60) / (MAX_IMMUTABLE_DATA_SIZE_IN_BYTES * 100 / RATE) = (OVERCHARGED_TIMEOUT_SECS + 60) * RATE / (MAX_IMMUTABLE_DATA_SIZE_IN_BYTES * 100)
which will always less than max_overcharged_entries ?

i think max_overcharged_entries shall be : OVERCHARGED_TIMEOUT_SECS / wait_millis  ?